### PR TITLE
Change some methods to use positional-only marker

### PR DESCRIPTION
**Pull request by bryanforbes**
## Summary

Changes some arguments in public methods to positional-only. A few of these (like the changes to `Client.dispatch()` and `BotBase.dispatch()`) fix incompatible override issues.

## Checklist

- [x] If code changes were made then they have been tested.
    - [x] I have updated the documentation to reflect the changes.
- [ ] This PR fixes an issue.
- [ ] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

**Comment by bryanforbes** [comment]
Thanks, @jack1142. I've added the changes to those sections.

File: discord/ext/commands/bot.py [hunk 1]
```diff
@@ -92,10 +92,18 @@
...
 
     These are meant to be passed into the :attr:`.Bot.command_prefix` attribute.
+
+        .. versionchanged:: 2.0
+
+            ``bot`` parameter is now positional-only.
+
+        .. versionchanged:: 2.0
+
+            ``msg`` parameter is now positional-only.
```
Comment (Rapptz) [inline discord/ext/commands/bot.py:106]:
```suggestion
        .. versionchanged:: 2.0

            ``bot`` and ``msg`` parameters arenow positional-only.
```

File: discord/ext/commands/bot.py [hunk 2]
```diff
@@ -217,6 +225,14 @@ async def on_command_error(self, context: Context[BotT], exception: errors.Comma
...
 
         This only fires if you do not specify any listeners for command error.
+
+        .. versionchanged:: 2.0
+
+            ``context`` parameter is now positional-only.
+
+        .. versionchanged:: 2.0
+
+            ``exception`` parameter is now positional-only.
```
Comment (Rapptz) [inline discord/ext/commands/bot.py:235]:
```suggestion
        .. versionchanged:: 2.0

            ``context`` and ``exception`` parameters are now positional-only.
```

File: discord/ext/commands/bot.py [hunk 3]
```diff
@@ -516,7 +559,7 @@ def remove_listener(self, func: CoroFunc, name: str = MISSING) -> None:
             except ValueError:
                 pass
 
-    def listen(self, name: str = MISSING) -> Callable[[CFT], CFT]:
+    def listen(self, name: str = MISSING, /) -> Callable[[CFT], CFT]:
```
Comment (Rapptz) [inline discord/ext/commands/bot.py:562]: I think this one is fine as a dual keyword/positional.

File: discord/ext/commands/bot.py [hunk 4]
```diff
@@ -816,7 +863,7 @@ def _resolve_name(self, name: str, package: Optional[str]) -> str:
         except ImportError:
             raise errors.ExtensionNotFound(name)
 
-    async def load_extension(self, name: str, *, package: Optional[str] = None) -> None:
+    async def load_extension(self, name: str, /, *, package: Optional[str] = None) -> None:
```
Comment (Rapptz) [inline discord/ext/commands/bot.py:866]: I think this one is fine as a dual keyword/positional.

File: discord/ext/commands/bot.py [hunk 5]
```diff
@@ -869,7 +920,7 @@ async def load_extension(self, name: str, *, package: Optional[str] = None) -> N
 
         await self._load_from_module_spec(spec, name)
 
-    async def unload_extension(self, name: str, *, package: Optional[str] = None) -> None:
+    async def unload_extension(self, name: str, /, *, package: Optional[str] = None) -> None:
```
Comment (Rapptz) [inline discord/ext/commands/bot.py:923]: I think this one is fine as a dual keyword/positional.

File: discord/ext/commands/bot.py [hunk 6]
```diff
@@ -916,14 +971,18 @@ async def unload_extension(self, name: str, *, package: Optional[str] = None) ->
         await self._remove_module_references(lib.__name__)
         await self._call_module_finalizers(lib, name)
 
-    async def reload_extension(self, name: str, *, package: Optional[str] = None) -> None:
+    async def reload_extension(self, name: str, /, *, package: Optional[str] = None) -> None:
```
Comment (Rapptz) [inline discord/ext/commands/bot.py:974]: I think this one is fine as a dual keyword/positional.

File: discord/ext/commands/help.py [hunk 7]
```diff
@@ -630,6 +663,14 @@ async def on_help_command_error(self, ctx: ContextT, error: CommandError) -> Non
         By default this method does nothing and just propagates to the default
         error handlers.
 
+        .. versionchanged:: 2.0
+
+            ``ctx`` parameter is now positional-only.
+
+        .. versionchanged:: 2.0
+
+            ``error`` parameter is now positional-only.
```
Comment (Rapptz) [inline discord/ext/commands/help.py:672]:
```suggestion
        .. versionchanged:: 2.0

            ``ctx`` and ``error`` parameters are now positional-only.
```

File: docs/migrating.rst [hunk 8]
```diff
@@ -1247,14 +1251,97 @@ define their type hints more accurately.
...
+- ``func`` in :meth:`ext.commands.Bot.check`
 - ``func`` in :meth:`ext.commands.Bot.add_check`
 - ``func`` in :meth:`ext.commands.Bot.remove_check`
+- ``func`` in :meth:`ext.commands.Bot.check_once`
+- ``ctx`` in :meth:`ext.commands.Bot.can_run`
+- ``func`` in :meth:`ext.commands.Bot.add_listener`
+- ``func`` in :meth:`ext.commands.Bot.remove_listener`
+- ``name`` in :meth:`ext.commands.Bot.load_extension`
+- ``name`` in :meth:`ext.commands.Bot.unload_extension`
+- ``name`` in :meth:`ext.commands.Bot.reload_extension`
```
Comment (Rapptz) [inline docs/migrating.rst:1332]:
```suggestion
```

File: docs/migrating.rst [hunk 9]
```diff
@@ -1247,14 +1251,97 @@ define their type hints more accurately.
...
+Parameters in the following methods are now all positional-only:
+
+- :meth:`ext.commands.when_mentioned`
+- :meth:`ext.commands.Bot.on_command_error`
+- :meth:`ext.commands.Bot.check`
+- :meth:`ext.commands.Bot.check_once`
+- :meth:`ext.commands.Bot.is_owner`
+- :meth:`ext.commands.Bot.before_invoke`
+- :meth:`ext.commands.Bot.after_invoke`
+- :meth:`ext.commands.Bot.listen`
```
Comment (Rapptz) [inline docs/migrating.rst:1263]:
```suggestion
```

File: discord/ext/commands/help.py [hunk 10]
```diff
@@ -777,6 +834,14 @@ async def prepare_help_command(self, ctx: ContextT, command: Optional[str] = Non
             This is called *inside* the help command callback body. So all
             the usual rules that happen inside apply here as well.
 
+        .. versionchanged:: 2.0
+
+            ``ctx`` parameter is now positional-only.
+
+        .. versionchanged:: 2.0
+
+            ``command`` parameter is now positional-only.
```
Comment (Rapptz) [inline discord/ext/commands/help.py:843]:
```suggestion
        .. versionchanged:: 2.0

            ``ctx`` and ``command`` parameters are now positional-only.
```

File: discord/ext/commands/help.py [hunk 11]
```diff
@@ -1143,14 +1225,22 @@ def get_ending_note(self) -> str:
...
         The default implementation is a bold underline heading followed
         by commands separated by an EN SPACE (U+2002) in the next line.
 
+        .. versionchanged:: 2.0
+
+            ``commands`` parameter is now positional-only.
+
+        .. versionchanged:: 2.0
+
+            ``heading`` parameter is now positional-only.
```
Comment (Rapptz) [inline discord/ext/commands/help.py:1242]:
```suggestion
        .. versionchanged:: 2.0

            ``commands`` and ``heading`` parameters are now positional-only.
```

**Review by Rapptz — CHANGES_REQUESTED** [review]
Thanks. Most of these I think are fair.

File: discord/ext/commands/bot.py (hunk 3 above)
Comment (bryanforbes) [reply discord/ext/commands/bot.py:562]:
```suggestion
    def listen(self, name: str = MISSING) -> Callable[[CFT], CFT]:
```

File: discord/ext/commands/bot.py (hunk 4 above)
Comment (bryanforbes) [reply discord/ext/commands/bot.py:866]:
```suggestion
    async def load_extension(self, name: str, *, package: Optional[str] = None) -> None:
```

File: discord/ext/commands/bot.py (hunk 5 above)
Comment (bryanforbes) [reply discord/ext/commands/bot.py:923]:
```suggestion
    async def unload_extension(self, name: str, *, package: Optional[str] = None) -> None:
```

File: discord/ext/commands/bot.py (hunk 6 above)
Comment (bryanforbes) [reply discord/ext/commands/bot.py:974]:
```suggestion
    async def reload_extension(self, name: str, *, package: Optional[str] = None) -> None:
```

File: discord/ext/commands/bot.py [hunk 12]
```diff
@@ -540,6 +583,10 @@ async def my_message(message):
 
         Would print one and two in an unspecified order.
 
+        .. versionchanged:: 2.0
+
+            ``name`` parameter is now positional-only.
+
```
Comment (bryanforbes) [inline discord/ext/commands/bot.py:589]:
```suggestion
```

File: discord/ext/commands/bot.py [hunk 13]
```diff
@@ -832,6 +879,10 @@ async def load_extension(self, name: str, *, package: Optional[str] = None) -> N
 
             This method is now a :term:`coroutine`.
 
+        .. versionchanged:: 2.0
+
+            ``name`` parameter is now positional-only.
+
```
Comment (bryanforbes) [inline discord/ext/commands/bot.py:885]:
```suggestion
```

File: discord/ext/commands/bot.py [hunk 14]
```diff
@@ -886,6 +937,10 @@ async def unload_extension(self, name: str, *, package: Optional[str] = None) ->
...
```
Comment (bryanforbes) [inline discord/ext/commands/bot.py:943]:
```suggestion
```

File: discord/ext/commands/bot.py [hunk 15]
```diff
@@ -916,14 +971,18 @@ async def unload_extension(self, name: str, *, package: Optional[str] = None) ->
...
 
         This replaces the extension with the same extension, only refreshed. This is
         equivalent to a :meth:`unload_extension` followed by a :meth:`load_extension`
         except done in an atomic way. That is, if an operation fails mid-reload then
         the bot will roll-back to the prior working state.
 
+        .. versionchanged:: 2.0
+
+            ``name`` parameter is now positional-only.
+
```
Comment (bryanforbes) [inline discord/ext/commands/bot.py:985]:
```suggestion
```

File: discord/ext/commands/help.py [hunk 16]
```diff
@@ -484,6 +496,14 @@ def subcommand_not_found(self, command: Command[Any, ..., Any], string: str) ->
...
             - If the ``command`` parameter has subcommands but not one named ``string``.
 
+        .. versionchanged:: 2.0
+
+            ``command`` parameter is now positional-only.
+
+        .. versionchanged:: 2.0
+
+            ``string`` parameter is now positional-only.
+
```
Comment (bryanforbes) [inline discord/ext/commands/help.py:506]:
```suggestion
        .. versionchanged:: 2.0

            ``command`` and ``string`` parameters are now positional-only.

```

**Review by Rapptz — APPROVED** [review]
Thanks!

**Pull request merged**